lib/AST/ASTPrinter.cpp

```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
     return T->hasSimpleTypeRepr();
   }
 
+  /// Add all Clang modules that have been imported into `Mod` to
+  /// `clangModules`.
+  ///
+  /// This also takes into account any modules that are imported transitively
+  /// through public (`@_exported`) imports.
+  /// `clangModules` is a map from each imported Clang module to the
+  /// corresponding Swift module.
+  static void getImportedClangModules(
+      ModuleDecl *Mod, ModuleDecl::ImportFilter filter,
+      llvm::DenseMap<const clang::Module *, ModuleDecl *> &clangModules) {
+    SmallVector<ModuleDecl::ImportedModule, 4> imports;
+    Mod->getImportedModules(imports, filter);
+    for (const auto &import : imports) {
+      if (const clang::Module *clangModule =
+              import.importedModule->findUnderlyingClangModule())
+        clangModules[clangModule] = import.importedModule;
+      getImportedClangModules(import.importedModule,
+                              ModuleDecl::ImportFilterKind::Public,
+                              clangModules);
+    }
+  }
+
+  /// Returns all Clang modules that are visible from `Options.CurrentModule`.
+  /// The returned map associates each visible Clang module with the
+  /// corresponding Swift module.
+  const llvm::DenseMap<const clang::Module *, ModuleDecl *> &
+  getVisibleClangModules() {
+    assert(Options.CurrentModule &&
+           "CurrentModule needs to be set to determine imported Clang modules");
+    if (!VisibleClangModules) {
+      VisibleClangModules.emplace();
+      ModuleDecl::ImportFilter Filter = ModuleDecl::ImportFilterKind::Public;
+      Filter |= ModuleDecl::ImportFilterKind::Private;
+      getImportedClangModules(Options.CurrentModule, Filter,
+                              *VisibleClangModules);
+    }
+    return *VisibleClangModules;
+  }
+
   template <typename T>
   void printModuleContext(T *Ty) {
     FileUnit *File = cast<FileUnit>(Ty->getDecl()->getModuleScopeContext());
     ModuleDecl *Mod = File->getParentModule();
+    std::string ExportedModuleName = File->getExportedModuleName();
+
+    // Clang declarations need special treatment: Multiple Clang modules can
+    // contain the same declarations from a textually included header, but not
+    // all of these modules may be visible. We therefore need to make sure we
+    // choose a module that is visible from the current module. This is
+    // obviously only possible, however, if we know what the current module is.
```

Comment:
```suggestion
    // choose a module that is visible from the current module. This is
    // possible to do only if we know what the current module is.
```

Reply:
Done.

Additionally, I've left out the "to do", as that seemed redundant.